File: crypto/bcrypt.go
package crypto

import (
	"github.com/LonelyPale/goutils"
	"github.com/LonelyPale/goutils/crypto/sha512"
)

const (
	DefaultLoopNumber = 2048
)

// 慢哈希加盐
func Bcrypt(data, salt []byte, loops ...int) ([]byte, error) {
	var loop int
	if len(loops) > 0 && loops[0] > 0 {
		loop = loops[0]
	} else {
		loop = DefaultLoopNumber
	}

	//1 正向哈希
	hash, err := sha512.Hash(data)
	if err != nil {
		return nil, err
	}

	//逆序盐
	saltReverse := goutils.ReverseByte(append([]byte{}, salt...))

	//2 前置盐
	salt, err = sha512.Hash(goutils.MergeSliceByte(salt, hash))
	if err != nil {
		return nil, err
	}

	//3 逆向哈希
	hashReverse, err := sha512.Hash(goutils.ReverseByte(data))
	if err != nil {
		return nil, err
	}
	goutils.ReverseByte(hashReverse)

	//4 后置盐
	saltReverse, err = sha512.Hash(goutils.MergeSliceByte(hashReverse, saltReverse))
	if err != nil {
		return nil, err
	}
	goutils.ReverseByte(saltReverse)

	//5 合并哈希
	newhash := make([]byte, 0)
	newhash = append(newhash, salt...)
	for i := range hash {
		newhash = append(newhash, hash[i])
		newhash = append(newhash, hashReverse[i])
	}
	newhash = append(newhash, saltReverse...)
	newhash, err = sha512.Hash(newhash)
	if err != nil {
		return nil, err
	}

	//6 重复哈希
	for i := 0; i < loop; i++ {
		if i%2 != 0 {
			goutils.ReverseByte(newhash)
		}

		newhash, err = sha512.Hash(newhash)
		if err != nil {
			return nil, err
		}
	}

	//7 重组哈希 4231是1、2、3、4能组成的最大素数
	b1 := make([]byte, 16)
	b2 := make([]byte, 16)
	b3 := make([]byte, 16)
	b4 := make([]byte, 16)
	for i := 0; i < 16; i++ {
		b1[i] = newhash[i*4]
		b2[i] = newhash[i*4+1]
		b3[i] = newhash[i*4+2]
		b4[i] = newhash[i*4+3]
	}
	a1 := b4
	a2 := goutils.ReverseByte(b2)
	a3 := goutils.ReverseByte(b3)
	a4 := b1
	newhash = goutils.MergeSliceByte(a1, a2, a3, a4)

	newhash, err = sha512.Hash(append(newhash, []byte("Bcrypt")...))
	if err != nil {
		return nil, err
	}

	return newhash, nil
}

func BcryptSimple(data, salt []byte, loops ...int) ([]byte, error) {
	var loop int
	if len(loops) > 0 && loops[0] > 0 {
		loop = loops[0]
	} else {
		loop = DefaultLoopNumber
	}

	hash, err := sha512.Hash(data)
	if err != nil {
		return nil, err
	}

	salt, err = sha512.Hash(salt)
	if err != nil {
		return nil, err
	}

	hash, err = sha512.Hash(append(hash, salt...))
	if err != nil {
		return nil, err
	}

	tempHash := hash
	tempSalt := salt
	hashs := make([]byte, 0)
	for i := 0; i < 10; i++ {
		tempHash, err = sha512.Hash(append(tempHash, byte(i)))
		if err != nil {
			return nil, err
		}

		tempSalt, err = sha512.Hash(append(tempSalt, byte(i)))
		if err != nil {
			return nil, err
		}

		tempHash, err = sha512.Hash(append(tempHash, tempSalt...))
		if err != nil {
			return nil, err
		}

		hashs = append(hashs, tempHash...)
	}

	hash, err = sha512.Hash(hashs)
	if err != nil {
		return nil, err
	}

	idxHash := len(hash) / 2
	hash1 := append([]byte{}, hash[:idxHash]...)
	hash2 := append([]byte{}, hash[idxHash:]...)

	idxSalt := len(salt) / 2
	salt1 := salt[:idxSalt]
	salt2 := salt[idxSalt:]

	for i := 0; i < loop; i++ {
		hash1, err = sha512.Hash(append(hash1, salt1...))
		if err != nil {
			return nil, err
		}

		hash2, err = sha512.Hash(append(hash2, salt2...))
		if err != nil {
			return nil, err
		}
	}

	hash, err = sha512.Hash(append(hash1, hash2...))
	if err != nil {
		return nil, err
	}

	hash, err = sha512.Hash(append(hash, []byte("BcryptSimple")...))
	if err != nil {
		return nil, err
	}

	return hash, nil
}

File: crypto/bcrypt.ts
import {enc, lib, SHA512} from "crypto-js";

export type WordArray = lib.WordArray;

//
export function BcryptSimple(data: WordArray, salt: WordArray, loops: number): WordArray {
    let hash = SHA512(data);
    // eslint-disable-next-line no-param-reassign
    salt = SHA512(salt);
    hash = SHA512(hash.concat(salt));

    let tempHash = hash.clone();
    let tempSalt = salt.clone();
    const hashs: WordArray = lib.WordArray.create();
    for (let i = 0; i < 10; i += 1) {
        // eslint-disable-next-line no-bitwise
        tempHash = SHA512(tempHash.concat(lib.WordArray.create([i << 24], 1)));

        // eslint-disable-next-line no-bitwise
        tempSalt = SHA512(tempSalt.concat(lib.WordArray.create([i << 24], 1)));

        tempHash = SHA512(tempHash.concat(tempSalt));
        hashs.concat(tempHash);
    }

    hash = SHA512(hashs);

    const idxHash = hash.sigBytes / 4 / 2;
    let hash1 = lib.WordArray.create(hash.words.slice(0, idxHash));
    let hash2 = lib.WordArray.create(hash.words.slice(idxHash));

    const idxSalt = salt.sigBytes / 4 / 2;
    const salt1 = lib.WordArray.create(salt.words.slice(0, idxSalt));
    const salt2 = lib.WordArray.create(salt.words.slice(idxSalt));

    hash1.clamp();
    hash2.clamp();

    for (let i = 0; i < loops; i += 1) {
        hash1 = SHA512(hash1.concat(salt1));
        hash2 = SHA512(hash2.concat(salt2));
    }

    hash = SHA512(hash1.concat(hash2));
    hash = SHA512(hash.concat(enc.Utf8.parse("BcryptSimple")));

    return hash;
}

/**/
function Bcrypt(data: WordArray, salt: WordArray, loops: number): WordArray {
    const hash = SHA512(data);
    let saltReverse = salt.clone();
    reverse(saltReverse);

    // eslint-disable-next-line no-param-reassign
    salt = SHA512(salt.concat(hash));
    const hashReverse = SHA512(reverse(data));
    reverse(hashReverse);

    saltReverse = SHA512(hashReverse.clone().concat(saltReverse));
    reverse(saltReverse);

    let newhash = lib.WordArray.create();
    newhash.concat(salt);
    for (let i = 0; i < hash.sigBytes; i += 1) {
        const b1 = getByteFromWords(hash.words, i);
        const b2 = getByteFromWords(hashReverse.words, i);
        // eslint-disable-next-line no-bitwise
        newhash.concat(lib.WordArray.create([b1 << 24], 1));
        // eslint-disable-next-line no-bitwise
        newhash.concat(lib.WordArray.create([b2 << 24], 1));
    }
    newhash.concat(saltReverse);
    newhash = SHA512(newhash);

    for (let i = 0; i < loops; i += 1) {
        if (i % 2 !== 0) {
            reverse(newhash);
        }
        newhash = SHA512(newhash);
    }

    const b1 = lib.WordArray.create();
    const b2 = lib.WordArray.create();
    const b3 = lib.WordArray.create();
    const b4 = lib.WordArray.create();
    for (let i = 0; i < 16; i += 1) {
        const byte1 = getByteFromWords(newhash.words, i * 4);
        const byte2 = getByteFromWords(newhash.words, i * 4 + 1);
        const byte3 = getByteFromWords(newhash.words, i * 4 + 2);
        const byte4 = getByteFromWords(newhash.words, i * 4 + 3);
        // eslint-disable-next-line no-bitwise
        b1.concat(lib.WordArray.create([byte1 << 24], 1));
        // eslint-disable-next-line no-bitwise
        b2.concat(lib.WordArray.create([byte2 << 24], 1));
        // eslint-disable-next-line no-bitwise
        b3.concat(lib.WordArray.create([byte3 << 24], 1));
        // eslint-disable-next-line no-bitwise
        b4.concat(lib.WordArray.create([byte4 << 24], 1));
    }
    const a1 = b4;
    const a2 = reverse(b2);
    const a3 = reverse(b3);
    const a4 = b1;
    newhash = lib.WordArray.create().concat(a1).concat(a2).concat(a3).concat(a4);

    newhash = SHA512(newhash.concat(enc.Utf8.parse("Bcrypt")));
    return newhash;
}


function reverse(data: WordArray): WordArray {
    const len = Math.floor(data.sigBytes / 2);
    for (let si = 0; si < len; si += 1) {
        const {words} = data;
        const ei = data.sigBytes - 1 - si;
        const start = getByteFromWords(words, si);
        const end = getByteFromWords(words, ei);
        setByteToWords(words, si, end);
        setByteToWords(words, ei, start);
    }
    return data;
}

function getByteFromWords(words: number[], indexByte: number): number {
    const i = Math.floor(indexByte / 4);
    const j = indexByte % 4;
    let temp: number = words[i];
    if (j === 0) {
        // eslint-disable-next-line no-bitwise
        temp >>>= 24;
    } else if (j === 1) {
        // eslint-disable-next-line no-bitwise
        temp = (temp & 0x00ff0000) >>> 16;
    } else if (j === 2) {
        // eslint-disable-next-line no-bitwise
        temp = (temp & 0x0000ff00) >>> 8;
    } else if (j === 3) {
        // eslint-disable-next-line no-bitwise
        temp &= 0x000000ff;
    }
    return temp;
}

function setByteToWords(words: number[], indexByte: number, num: number) {
    const i = Math.floor(indexByte / 4);
    const j = indexByte % 4;
    const twords = words;
    if (j === 0) {
        // eslint-disable-next-line no-bitwise
        twords[i] = (twords[i] & 0x00ffffff) | (num << 24);
    } else if (j === 1) {
        // eslint-disable-next-line no-bitwise
        twords[i] = (twords[i] & 0xff00ffff) | (num << 16);
    } else if (j === 2) {
        // eslint-disable-next-line no-bitwise
        twords[i] = (twords[i] & 0xffff00ff) | (num << 8);
    } else if (j === 3) {
        // eslint-disable-next-line no-bitwise
        twords[i] = (twords[i] & 0xffffff00) | num;
    }
}

const data = enc.Utf8.parse("hello world");
const key = enc.Utf8.parse("123");
console.log("BcryptSimple:", BcryptSimple(data, key, 2048).toString());
console.log("Bcrypt:", Bcrypt(data, key, 2048).toString());
